### ISSUE TITLE
Add ASR debugging facility

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#1384]
To get started, for the LLVM backend let's save the ASR in the `--show-asr --indent` form after each pass to a sequence of files. Then write a simple script that shows them side by side, one pass at a time, and I can browse through it. Possibly also colorful "diff" might be helpful.

**Comment by Shaikh-Ubaid** [comment]
>write a simple script that shows them side by side

How are we planning to show the outputs side by side? Like what would be the GUI for it?

**Comment by certik** [comment]
A terminal gui would be nice. Or an html page. We'll have to play with it, until we find something that we like. The purpose is to use it for Debugging, as well as to learn about what the ASR pass is doing.

**Comment by Shaikh-Ubaid** [comment]
Would this work?
![image](https://user-images.githubusercontent.com/43722035/222296929-dbefbd73-249e-44cc-993a-b68759db7ca7.png)

 (I think `<` indicates removal, `>` indicates addition and `|` indicates modification).

We can call the `diff` from a python script using `os.system()`.

**Comment by certik** [comment]
Yes, it would.

**Comment by certik** [comment]
And yes, I think we want to look at the ASR using the (to be written) Fortran and/or Python backend, as well as the `--show-asr --indent` backend. We'll start with the latter, since we already have it, and we'll need that anyway, to see all the information, but later the Fortran backend would be great to see an overview of what is going on.

**Comment by czgdp1807** [comment]
Quoting from our (@certik's and mine) discussions,

> What I would like to see is an option to apply passes cumulatively. We have a list of passes in pass_manager.h. Say print_arr is at 6th position. Then I want an option, "--pass=print_arr --cumulative". This would apply 6 passes till print_arr. I have felt the need for it in my previous debugging sessions.
> For now I comment out the list and compile the project again but a command line option would save that effort.

The above would be a great addition from my personal point of view.

**Comment by Pranavchiku** [comment]
> The above would be a great addition from my personal point of view.

Yeah, I have also felt the need for the same while debugging ASR verify errors or some segmentation fault errors, and many a time I come across small fix in one or two passes only, so having this idea will make debugging easier.
